ships/0036-runAs-for-supporting-steps.md

```diff
@@ -0,0 +1,122 @@
+
+1. The source code on the file system is owned by this user and group.
+2. As long as we have creds-init enabled, the `.docker` directory in `/tekton/home` is owned by this user and group.
+
+Build strategy authors can also define a securityContext for each step where they can define a runAsUser and runAsGroup.
+
+Different sample build strategies need to run as different users and groups. For example:
+
+- [Kaniko is running as user 0](https://github.com/shipwright-io/build/blob/v0.11.0/samples/buildstrategy/kaniko/buildstrategy_kaniko_cr.yaml#L15) (root) due to the way is building the container image.
+- [BuildKit runs as user 1000 and group 1000](https://github.com/shipwright-io/build/blob/v0.11.0/samples/buildstrategy/buildkit/buildstrategy_buildkit_cr.yaml#L58-L59)
+- [Paketo Buildpacks today run as user 1000 and group 1000](https://github.com/shipwright-io/build/blob/v0.11.0/samples/buildstrategy/buildpacks-v3/buildstrategy_buildpacks-v3_cr.yaml#L40-L41). That's only the case as long as we are using the Bionic builder image there. For Jammy, the Paketo community actually updated the stack to user 1001 and group 1000 at build time. We will need to adapt this in our sample strategy when we switch to Jammy there.
+- [ko runs as user 1000 and group 1000](https://github.com/shipwright-io/build/blob/v0.11.0/samples/buildstrategy/ko/buildstrategy_ko_cr.yaml#L55-L56)
+
+All sample build strategies that run as non-root currently contain a [prepare step](https://github.com/shipwright-io/build/blob/v0.11.0/samples/buildstrategy/ko/buildstrategy_ko_cr.yaml#L29-L49). This step ensures that `/tekton/home` is owned by the build strategy runAsUser/runAsGroup so that creds-init also works there IF the source step runs as a different user. With today's default configuration, this is a no-op because all source steps by default run as user=1000 and group=1000 which is the same non-root user of all our sample build strategies. It is only there in case somebody uses a different user, or group for the source step. Though, I think we really never tested this and are currently having an issue because we do not `chown` the source directory as well. Basically if the source step is reconfigured as something else, then the runAsUser/runAsGroup of our sample build strategies may not have full access insight the source directory which can lead to failures.
+
+To improve this, I am suggesting to change our logic to run the supporting steps as the user that is used in the build strategy IF all build strategy steps are defined to run as the same user. This means that all steps will run as the same user. The same logic applies to the group.
```

Comment:
Can you make a distinction on the following:

- When you say all steps are defined to run as the same user, do you mean that all containers uses the same `securityContext` definition, or that the pod `securityContext` is defined based on the strategy step. The difference here, is that one can override any step(container) securityContext eventually.

Reply:
My plan is to not touch the pod's security context and do everything on the containers instead. Basically as described under "BuildRun reconciler changes" = check the runAsUsers of all strategy steps. If they are always defined and the same, then use the this value for the supporting steps that we manage. Same for the group.

The pod's security context would be configurable through the [TaskRun's pod template](https://tekton.dev/docs/pipelines/taskruns/#specifying-a-pod-template). Would also work, but given we don't set this (the pod template itself) today, I am not considering this option atm. It might be a way to also run Tekton's initContainers as this user. So far I don't think we need that simply because everything that I tried worked without that I changed anything for it.

> The difference here, is that one can override any step(container) securityContext eventually

Correct. Though, the goal here is to run everything as the same user if the strategy does this as well.

Let me know if you want me to add a section to alternatives.

Reply:
That would be a good candidate for Alternatives. I do still have a concern on for example Kaniko, where we use user ID 0. Not comfortable setting the same across all containers. Specially once we adopt the additional features with the managed-push layer, which means new tools we don't own(binaries). What's your take on this?

Reply:
Understand. I guess we need to distinguish two scenarios when we run as root.

1. Kaniko. If you decide to use Kaniko, then you decide that the build user is allowed to run arbitrary commands as root based on the Dockerfile.
2. Not Kaniko. We don't have a strategy for what I think of. But basically a scenario where the commands that run are under full control of the build strategy author but still must run as root for whatever reason. Here, I understand the concern.

I'd still would want to solve it with just docs these days.